### PR TITLE
HOSTEDCP-917: Add publicAndPrivate <-> Private e2e test

### PR DESCRIPTION
**Pull request by muraee**
**What this PR does / why we need it**:
Validates switching a cluster from `PublicAndPrivate` to `Private` and vice versa to be working .

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #[HOSTEDCP-917](https://issues.redhat.com/browse/HOSTEDCP-917)

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Review by bryan-cox — APPROVED** [review]
/lgtm

File: test/e2e/create_cluster_test.go [hunk 1]
```diff
@@ -217,24 +182,32 @@ func TestCreateClusterProxy(t *testing.T) {
...
 	g.Expect(err).NotTo(HaveOccurred(), "failed to get hostedcluster")
 
-	e2eutil.EnsureNodeCountMatchesNodePoolReplicas(t, testContext, client, guestClient, hostedCluster.Namespace)
+	e2eutil.EnsureNodeCountMatchesNodePoolReplicas(t, ctx, client, guestClient, hostedCluster.Namespace)
 	e2eutil.EnsureNoCrashingPods(t, ctx, client, hostedCluster)
 	e2eutil.EnsureNodeCommunication(t, ctx, client, hostedCluster)
 }
+
+func validatePrivateCluster(t *testing.T, ctx context.Context, client crclient.Client, hostedCluster *hyperv1.HostedCluster) {
+	g := NewWithT(t)
```
Comment (enxebre) [inline test/e2e/create_cluster_test.go:198]: can we please include here validations for known expectations:
E.g. the format of the hc.status.controlplaneendpoint is as expected then "public" URL and not local?

File: test/e2e/create_cluster_test.go [hunk 2]
```diff
@@ -217,24 +183,41 @@ func TestCreateClusterProxy(t *testing.T) {
...
+	// Ensure NodePools have all Nodes ready.
+	e2eutil.WaitForNodePoolDesiredNodes(t, ctx, client, hostedCluster)
+
+	// Wait for the rollout to be complete
+	t.Logf("Waiting for cluster rollout. Image: %s", globalOpts.LatestReleaseImage)
+	e2eutil.WaitForImageRollout(t, testContext, client, hostedCluster, globalOpts.LatestReleaseImage)
+
+	serviceStrategy := util.ServicePublishingStrategyByTypeByHC(hostedCluster, hyperv1.APIServer)
+	g.Expect(serviceStrategy).ToNot(BeNil())
+	if serviceStrategy.Type == hyperv1.Route && serviceStrategy.Route != nil {
```
Comment (enxebre) [inline test/e2e/create_cluster_test.go:213]: nit: I'd fail here if `serviceStrategy.Type != hyperv1.Route`

File: test/e2e/create_cluster_test.go [hunk 3]
```diff
@@ -217,24 +183,41 @@ func TestCreateClusterProxy(t *testing.T) {
...
+		g.Expect(hostedCluster.Status.ControlPlaneEndpoint.Host).To(Equal(serviceStrategy.Route.Hostname))
```
Comment (enxebre) [inline test/e2e/create_cluster_test.go:214]: this validation is actually not specific not private cluster can we include in both public and private?

**Comment by muraee** [comment]
/hold
waiting on https://github.com/openshift/hypershift/pull/2389

**Comment by enxebre** [comment]
/hold 
to make sure we don't introduce flakiness

File: test/e2e/util/node.go [hunk 4]
```diff
@@ -24,10 +26,19 @@ func EnsureNodeCommunication(t *testing.T, ctx context.Context, client crclient.
 		g.Expect(err).NotTo(HaveOccurred(), "couldn't load guest kubeconfig")
 		guestClient := kubeclient.NewForConfigOrDie(guestConfig)
 
-		podList, err := guestClient.CoreV1().Pods("kube-system").List(ctx, metav1.ListOptions{LabelSelector: "app=konnectivity-agent"})
-		g.Expect(err).NotTo(HaveOccurred())
-		g.Expect(podList.Items).ToNot(BeEmpty())
-		_, err = guestClient.CoreV1().Pods("kube-system").GetLogs(podList.Items[0].Name, &corev1.PodLogOptions{Container: "konnectivity-agent"}).DoRaw(ctx)
+		err = wait.PollImmediateWithContext(ctx, 10*time.Second, 5*time.Minute, func(ctx context.Context) (done bool, err error) {
```
Comment (enxebre) [inline test/e2e/util/node.go:29]: how is this change related to this PR?

File: test/e2e/util/util.go [hunk 5]
```diff
@@ -139,7 +143,7 @@ func WaitForGuestClient(t *testing.T, ctx context.Context, client crclient.Clien
 
 	t.Logf("Waiting for a successful connection to the guest apiserver")
 	var guestClient crclient.Client
-	waitForGuestClientCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	waitForGuestClientCtx, cancel := context.WithTimeout(ctx, 15*time.Minute)
```
Comment (enxebre) [inline test/e2e/util/util.go:146]: why is this increasing?

File: test/e2e/util/node.go (hunk 4 above)
Comment (muraee) [reply test/e2e/util/node.go:29]: some runs failed to get the `konnectivity-agent`  pod immediately after switching from private to public. This simply adds a retry to solve that.

File: test/e2e/util/util.go (hunk 5 above)
Comment (muraee) [reply test/e2e/util/util.go:146]: kubeconfig takes longer to report available after switching from private to public

File: test/e2e/util/node.go (hunk 4 above)
Comment (enxebre) [reply test/e2e/util/node.go:29]: Do we know what was causing the gail to get?
Can we please add a comment elaborating the above so the change is justified and we understand where this is coming from in future?

File: test/e2e/util/util.go (hunk 5 above)
Comment (enxebre) [reply test/e2e/util/util.go:146]: Please let's add a comment that justifies the change for future context.
